src/jdk.jpackage/windows/native/common/MsiDb.cpp

```diff
@@ -0,0 +1,282 @@
+        const auto status = MsiViewClose(hView);
+        if (status != ERROR_SUCCESS) {
+            LOG_WARNING(tstrings::any() << "MsiViewClose("
+                                << hView << ") failed with error=" << status);
+            return;
+        }
+        closeMSIHANDLE(hView);
+    }
+}
+
+
+namespace {
+UniqueMSIHANDLE openDatabase(const tstring& msiPath) {
+    MSIHANDLE h = 0;
+    const UINT status = MsiOpenDatabase(msiPath.c_str(),
+                                                    MSIDBOPEN_READONLY, &h);
+    if (status != ERROR_SUCCESS) {
+        JP_THROW(Error(tstrings::any()
+                                << "MsiOpenDatabase(" << msiPath
+                                << ", MSIDBOPEN_READONLY) failed", status));
+    }
+    return UniqueMSIHANDLE(h);
+}
+
+} // namespace
+
+Database::Database(const Guid& productCode):
+        msiPath(getProductInfo(productCode, INSTALLPROPERTY_LOCALPACKAGE)),
+                                            dbHandle(openDatabase(msiPath)) {
+}
+
+
+Database::Database(const tstring& msiPath): msiPath(msiPath),
+                                            dbHandle(openDatabase(msiPath)) {
+}
+
+
+tstring Database::getProperty(const tstring& name) const {
+    // Query value of a property with the given name from 'Property' MSI table.
+    const tstring sqlQuery = (tstrings::any()
+                    << "SELECT Value FROM Property WHERE Property = '"
+                    << name << "'").tstr();
+
+    DatabaseView view(*this, sqlQuery);
+    const DatabaseRecord record(view);
+
+    // Data is stored in a record object. SQL query is constructed in a way
+    // this record object contains a single field.
+    // Verify record contains exactly one field.
+    if (record.getFieldCount() != 1) {
+        JP_THROW(Error(
+                    tstrings::any() << "record.getFieldCount(" << msiPath
+                                    << ", " << sqlQuery
+                                    << ") returned unexpected value",
+                    ERROR_SUCCESS));
+    }
+
+    // Field identifier. They start with 1, not from 0.
+    const unsigned field = 1;
+    return record.getString(field);
+}
+
+
+tstring Database::getProperty(const std::nothrow_t&, const tstring& name) const {
+    try {
+        return getProperty(name);
+    } catch (const NoMoreItemsError&) {
+    }
+    JP_CATCH_EXCEPTIONS;
+    return tstring();
+}
+
+
+DatabaseRecord::DatabaseRecord(unsigned fieldCount) {
+    MSIHANDLE h = MsiCreateRecord(fieldCount);
+    if (!h) {
+        JP_THROW(msi::Error(tstrings::any() << "MsiCreateRecord("
+                        << fieldCount << ") failed", ERROR_FUNCTION_FAILED));
+    }
+    handle = UniqueMSIHANDLE(h);
+}
+
+
+DatabaseRecord& DatabaseRecord::operator=(const DatabaseRecord& other) {
+    DatabaseRecord tmp(other);
+    std::swap(handle, tmp.handle);
+    return *this;
+}
+
+
+DatabaseRecord& DatabaseRecord::fetch(DatabaseView& view) {
+    *this = view.fetch();
```

Comment:
Another this modification and one more below.

Reply:
This is implicit call of class assignment operator. No crime :)